Homework 20

timepritchard opened the pull request:
I have refactored the code to replace the thread.sleep with an explicit wait, however, homework 19 is working but homeworks 18 and 17 are now not working and I'm not sure why?

Berichicko commented:
> I have refactored the code to replace the thread.sleep with an explicit wait, however, homework 19 is working but homeworks 18 and 17 are now not working and I'm not sure why?

Hi
You got this problem, please correct it. I will try to help you!
![image](https://github.com/user-attachments/assets/eaba2259-5afa-483b-8ef7-da433a0d0cd6)

**HINT**

1) How to analyze the results of tests:

Test status: 
![Test status_en](https://github.com/Test-Pro-Bootcamp/A53-QA-Automation-Framework/assets/104761482/bfa74efc-b45b-4ca0-8d89-f5224a51bc66)

	
1) Test error (red). This status is assigned to tests that caused an exception from the tested source code.

The tester needs to identify an issue in their code: Test error. This status is assigned to tests that trigger an exception in the tested source code.
In software testing, the "Test error" status is assigned to tests that trigger an exception in the tested source code. When a test case fails due to an unexpected exception being thrown during its execution, it is marked as a test error. This indicates that the code being tested has not handled the exception properly, and further investigation and debugging are required to address the issue.

This status indicates an error in the code and the tester must fix this error:

![Test error_en](https://github.com/Test-Pro-Bootcamp/A53-QA-Automation-Framework/assets/104761482/701a79a3-7c65-4a84-9180-e221c559ea02)

2) Test failed(yellow). If at least one test receives this status, then all its parents are marked as failed. (BUGS)
![Test failed_en](https://github.com/Test-Pro-Bootcamp/A53-QA-Automation-Framework/assets/104761482/0171eacb-5e3f-4e2a-9a35-e0e626260d1d)

**Step 1**

Homework17 

Create a playlist "Tim"

![image](https://github.com/user-attachments/assets/7d226d29-432c-49fa-9236-ccf3f5d80d6a)

**Step 2**

 **Use in test methods, the following logical split:   GIVEN – WHEN - THEN**
  
![image](https://github.com/user-attachments/assets/4c282431-8b39-452e-ac4d-f80dce6e78cb)

```
 public void clickPlay() {
    WebElement playButton = driver.findElement(By.cssSelector("span[data-testid='play-btn']"));
    Actions actions = new Actions(driver);
    try {
      actions.moveToElement(playButton).perform();
      actions.click().build().perform();
    } catch (TimeoutException e) {
      System.out.println("Standard click failed, trying JavaScript click");
      ((JavascriptExecutor) driver).executeScript("arguments[0].click();", playButton);
    }
  }
```

This clickPlay method is designed to click the "Play" button on a web page. It works like this:

1. Finds the "Play" button using a CSS selector span[data-testid='play-btn'].
2. Creates an Actions object to perform complex user interactions.
3. Tries to move the cursor to the button and click it using Actions. https://www.selenium.dev/documentation/webdriver/actions_api/mouse/
4. If a TimeoutException occurs, it will attempt to click the button using JavaScript. https://www.baeldung.com/java-selenium-javascript

The method includes both a standard way to click the button and a backup plan in case the first method fails.

```
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.testng.Assert;
import org.testng.annotations.Test;

public class Homework18 extends BaseTest {

  @Test
  public void playSong() {
//      GIVEN
    inputEmail("timothy.pritchard@testpro.io");
    inputPassword("q9RQ8fbN");
//      WHEN
    submitLogin();
    allSongs();
    selectSong();
    clickPlay();
//      THEN
    Assert.assertTrue(isDisplayedPlayingSong());
  }

  public void clickPlay() {
    WebElement playButton = driver.findElement(By.cssSelector("span[data-testid='play-btn']"));
    Actions actions = new Actions(driver);
    try {
      actions.moveToElement(playButton).perform();
      actions.click().build().perform();
    } catch (TimeoutException e) {
      System.out.println("Standard click failed, trying JavaScript click");
      ((JavascriptExecutor) driver).executeScript("arguments[0].click();", playButton);
    }
  }


  public boolean isDisplayedPlayingSong() {
    //WebElement songIsPlaying = driver.findElement(By.cssSelector("[data-testid = 'sound-bar-play']"));
    WebElement songIsPlaying = wait.until(ExpectedConditions.visibilityOfElementLocated
        (By.cssSelector("[data-testid = 'sound-bar-play']")));
    return songIsPlaying.isDisplayed();
  }


  public void selectSong() {
    //WebElement song = driver.findElement(By.cssSelector("tr.song-item"));
    WebElement song = wait.until(ExpectedConditions.visibilityOfElementLocated
        (By.cssSelector("tr.song-item")));
    song.click();
  }

  public void allSongs() {
    //WebElement allSongsTab = driver.findElement(By.cssSelector("a[href='#!/songs']"));
    WebElement allSongsTab = wait.until(ExpectedConditions.visibilityOfElementLocated
        (By.cssSelector("a[href='#!/songs']")));
    allSongsTab.click();
  }

}
```

timepritchard commented:
Hi, thanks for your help, I have updated with the click action in homework 18. What i don't understand is why was it working before I refactored the waits, and then after I refactored the waits it wasnt working and needed the action object?

Berichicko commented:
> Hi, thanks for your help, I have updated with the click action in homework 18. What i don't understand is why was it working before I refactored the waits, and then after I refactored the waits it wasnt working and needed the action object?

Hi

UI tests can be unstable, so they need to be continuously analyzed and maintained. An Automation QA engineer's job includes monitoring and fixing errors (exceptions) that occur in the tests.

For example

**Key Steps:**
Monitor Errors: Regularly check test logs to identify and understand the causes of errors.
Fix Tests: Update tests based on the errors found, improve locators, and add waits.
Refactor Code: Periodically improve the structure and readability of test code to make it more stable.
Use Reports: Set up automatic report generation to quickly find and analyze errors.
Work Together: Discuss test results with the team to quickly resolve issues.

![image](https://github.com/user-attachments/assets/fd9c4c89-a0fe-4156-9e9d-1b6f9a94124f)

Regularly analyzing and fixing errors in UI tests ensures their stability and reliability, which is a crucial part of an Automation QA engineer's job.
